src/dodal/common/beamlines/instantiation_behaviour.py

```diff
@@ -0,0 +1,124 @@
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from bluesky.run_engine import call_in_bluesky_event_loop
+from ophyd_async.core import DEFAULT_TIMEOUT
+from ophyd_async.core import Device as OphydV2Device
+
+from dodal.common.beamlines.beamline_utils import (
+    ACTIVE_DEVICES,
+)
+from dodal.utils import AnyDevice
+
+
+class XYZDetector(OphydV2Device):
+    def __init__(self, prefix: str, *args, **kwargs):
+        self.prefix = prefix
+        super().__init__(*args, **kwargs)
+
+    @property
+    def hints(self):
+        raise NotImplementedError
+
+
+@dataclass
+class DeviceInitializationConfig:
+    """
+    eager ones all connect at the beginning in blueapi startup. if they fail to connect we get an error
+    endpoint for what devices are failing
+    only matters in connect failure reporting behavior
+    lazy and eager are not about when connect but whether should be connected
+    """
+
+    eager: bool = True
+    set_name: bool = True
+    default_timeout_for_connect: float = DEFAULT_TIMEOUT
+    default_use_mock_at_connection: bool = False
+
+
+class DeviceInitializationController:
+    device: AnyDevice | None = None
+    config: DeviceInitializationConfig | None = None
+
+    def __init__(
+        self, config: DeviceInitializationConfig, factory: Callable[[], AnyDevice]
+    ) -> None:
+        self.factory = factory
+        self.config = config
+
+    # TODO right now the cache is in a global variable ACTIVE_DEVICES, that should change
+    def see_if_device_is_in_cache(self, name: str) -> AnyDevice | None:
+        d = ACTIVE_DEVICES.get(name)
+        assert isinstance(d, AnyDevice) or d is None
+        return d
+
+    def add_device_to_cache(self, device: AnyDevice) -> None:
+        ACTIVE_DEVICES[device.name] = device
```

Comment:
We still have the issue that `factory.__name__` doesn't always equal `device.name`. Solution is to put it in a dictionary keyed by factory rather than `device.name`. It can still be placed in `ACTIVE_DEVICES` if that is needed.

@callumforrester what is `ACTIVE_DEVICES` needed for?

Reply:
but factory name is same for all devices of the same type, right?

I have 3 slits at i22, slit2, slit3, slit3 - they all use the same SlitDevice class as the factory

Reply:
ACTIVE_DEVICES are consumed by blueapi 

https://github.com/DiamondLightSource/blueapi/pull/515

Reply:
I thought that blueapi only needed the factories, not the devices themselves...

Reply:
there is an endpoint which returns all the devices

Reply:
@callumforrester we have some lack of clarity, I think you might have the antidote information to that

Reply:
I'm not sure if I do, but I can try.

@coretl is referring to the function name, not the return type, so

```python
# Factory name is detector_1, device name is detector_1
def detector_1() -> MyDetector:
    return MyDetector(name="detector_1")

# Factory name is detector_2, device name is pilatus_2
def detector_2() -> MyDetector:
    return MyDetector(name="pilatus_2")
```

There is no guarantee that the function name and device name will match up. The question is, if you want to use the second device in a plan, should you pass `"detector_2"` or `"pilatus_2"` as a parameter? We should consider the case of nested devices too. Do we want `detector_2.exposure`, `pilatus_2.exposure` or `pilatus_2-exposure`. The latter is the name automatically assigned to the signal on device connection.

I think there are three options:
1. Key devices by name (`pilatus_1` in this case) and completely ignore the factory name in blueapi. If you look for a device called `my_device` you'll be out of luck. 
2. Key devices by name (`my_device` in this case) and completely ignore the device name in blueapi. `pilatus_1` would only appear in the documents in this case.
3. Raise an error, either in blueapi or dodal, if the names do not match up. Force them to be the same. This would be potentially annoying when writing devices but would make debugging easier. 

1 and 2 are workable, but include hidden knowledge and coupling, they would need to be well documented. 3 just includes hidden coupling, the knowledge is exposed to the user through an exception.

I _believe_ 2 is the current behaviour, so we should probably stick with that and raise a separate ticket if we think it should change?

I'm hoping someone else has other ideas, because I'm not a huge fan of any of these. @coretl originally wanted ophyd-async devices to be named after the variables to which they were assigned, but I believe @DominicOram had some objection to that...?

I'm also going to canvas @DiamondJoseph @olliesilvester @abbiemery and @evalott100 for opinions.

Reply:
How about 1, but the device name defaults to the factory name

Reply:
I thought the point of this issue was to make Device.name be injected and always equal to the DeviceFactoryFunction.__name__?

I think blueapi should only care about the device.name.

Reply:
I actually like keeping them the same and I would be open to the `name` not needing to be specified on the instantiation and just having it filled in by the factory decorator as the function name. Yes, it's magic, which I normally don't like but it's not too bad, basically just the decorator filling in an argument for you.

Reply:
what if we have multiple devices with the same factory name?

at i22 there are multiple slits

Reply:
They're functions in the same namespace, right? So they syntactically cannot have the same name.

In i22 they are `slits_1`, `slits_2` etc.

Reply:
ok I see

Reply:
https://github.com/DiamondLightSource/blueapi/issues/440#issue-2268699991

this is the relevant Blueapi PR
`Call dodal.beamlines.beamline_utils.get_device_factories() to get all the device factories and whether they are lazy or not`